docs/JLBP-14.md

```diff
@@ -51,15 +51,21 @@ cause the build to fail early when multiple versions are mixed. For example,
 `io.grpc:grpc-alts:1.25.0` and `io.grpc:grpc-core:1.23.0` cannot appear in 
 the classpath at the same time. If dependency mediation selects both
 `io.grpc:grpc-alts:1.25.0` and `io.grpc:grpc-core:1.23.0`, the build fails; and the developer must fix the conflict before they can proceed.
+
+Rapidly changing, pre-1.0 projects with unstable APIs can use hard
```

Comment:
Is it a problem for post-1.0 projects with internal unstable APIs to use hard requirements for intraproject dependencies?

Reply:
Yes, it's a problem for dependents no matter why or when you do it.

Reply:
How is it a problem? The only time it causes failures is when it *should* cause failures (i.e. when other means of detecting version mismatches have failed). The way grpc uses it, I don't think it has false positives.

Reply:
It's a problem for the same reasons hard dependencies are always a problem: they make builds brittle and lead to hard to diagnose and fix errors. gRPC has imposed extra cost on dependents by  using this technique.

Reply:
The technique moves obscure runtime failures into build time failures. It's the same effect as dependency convergence, except that it only fails at the correct time.

@ejona86 can you weigh in? I'm arguing in favor of grpc's usage of hard requirements for internal dependencies, @elharo is arguing against it.

Reply:
These shouldn't be obscure runtime failures either. They shouldn't be failures at all. Tight coupling between components causes problems for users. If the code can't be more decoupled and robust, then it shouldn't be in separate artifacts.

Reply:
This sort of stuff doesn't exist in a vacuum. I would agree such internal unstable APIs should be avoided.

But if gRPC can't use internal APIs then our options are to 1) bump major versions ~yearly (and sometimes more frequently since some changes are time sensitive) along with package renaming or 2) have larger artifacts with more dependencies (which actually doesn't work because Netty vs OkHttp wouldn't really fit with this model). (Or option (3), give up on performance.) Shading can help ease (2), but would cause us to use much more resources at runtime (like Threads) and only solves version pinning with grpc-core, not grpc-api. Remember that if it weren't for dependencies of the artifacts, they _would_ be all in one artifact. We don't _want_ them to be separate artifacts; they _need_ to be separate artifacts.

So you sort of need to pick your poison, and having to fight Maven to match the versions doesn't seem outrageous since you are _already_ fighting Maven to get upperBoundDeps which is similar complexity.

Reply:
Yes, I agree with this, and this agrees with the proposed changes. You have to pick your poison, but this is poison. It does impose real costs on dependents, and should only be done if it's the least bad option. I don't want to mislead library designers into believing that using hard dependencies for intraproject dependencies won't cause problems.